src/azure-cli-core/azure/cli/core/auth/identity.py

```diff
@@ -57,35 +66,53 @@ def __init__(self, authority, tenant_id=None, client_id=None, encrypt=False):
         self._token_cache_file = os.path.join(config_dir, "msal_token_cache")
         self._secret_file = os.path.join(config_dir, "service_principal_entries")
-        self._http_cache_file = os.path.join(config_dir, "msal_http_cache")
-
-        # Prepare HTTP cache.
-        # https://github.com/AzureAD/microsoft-authentication-library-for-python/pull/407
-        # if not Identity.http_cache:
-        #     Identity.http_cache = self._load_http_cache()
+        self._http_cache_file = os.path.join(config_dir, "msal_http_cache.bin")
 
+        # We make _msal_app_instance an instance attribute, instead of a class attribute,
+        # because MSAL apps can have different tenant IDs.
         self._msal_app_instance = None
-        # Store for Service principal credential persistence
-        self._msal_secret_store = ServicePrincipalStore(self._secret_file, self.encrypt)
-        self._msal_app_kwargs = {
+
+    @property
+    def _msal_app_kwargs(self):
+        """kwargs for creating UserCredential or ServicePrincipalCredential.
+        MSAL token cache and HTTP cache are lazily created.
+        """
+        if not Identity._msal_token_cache:
+            Identity._msal_token_cache = self._load_msal_token_cache()
+
+        if not Identity._msal_http_cache:
+            Identity._msal_http_cache = self._load_msal_http_cache()
+
+        return {
             "authority": self._msal_authority,
-            "token_cache": self._load_msal_cache()
-            # "http_cache": Identity.http_cache
+            "token_cache": Identity._msal_token_cache,
+            "http_cache": Identity._msal_http_cache
         }
 
-    def _load_msal_cache(self):
+    @property
+    def _msal_app(self):
+        """A PublicClientApplication instance for user login/logout.
+        The instance is lazily created.
+        """
+        if not self._msal_app_instance:
+            self._msal_app_instance = PublicClientApplication(self.client_id, **self._msal_app_kwargs)
+        return self._msal_app_instance
+
+    def _load_msal_token_cache(self):
         # Store for user token persistence
         cache = load_persisted_token_cache(self._token_cache_file, self.encrypt)
         return cache
 
-    def _load_http_cache(self):
+    def _load_msal_http_cache(self):
         import atexit
         import pickle
 
+        logger.debug("_load_msal_http_cache: %s", self._http_cache_file)
         try:
             with open(self._http_cache_file, 'rb') as f:
-                persisted_http_cache = pickle.load(f)  # Take a snapshot
-        except:  # pylint: disable=bare-except
+                persisted_http_cache = pickle.load(f)
+        except (pickle.UnpicklingError, FileNotFoundError) as ex:
+            logger.debug("Failed to load MSAL HTTP cache: %s", ex)
             persisted_http_cache = {}  # Ignore a non-exist or corrupted http_cache
         atexit.register(lambda: pickle.dump(
             # When exit, flush it back to the file.
```

Comment:
Yes it is effectively the same as [MSAL's recommended pattern](https://msal-python.readthedocs.io/en/latest/). Ship it!

Reply:
Well, I am a little bit confused. Haven't we decided not to catch `EOFError, AttributeError, ImportError, IndexError` (https://github.com/AzureAD/microsoft-authentication-library-for-python/pull/407#discussion_r727329068)? They still appear in https://msal-python.readthedocs.io/en/latest/#publicclientapplication

![image](https://user-images.githubusercontent.com/4003950/141037325-391923b0-44b8-4278-9819-d0fda3547ae4.png)

Reply:
You are right. I created a new PR in MSAL to remove them. https://github.com/AzureAD/microsoft-authentication-library-for-python/pull/439